### ISSUE TITLE
With dependency pruning activated, we do not detect bugs that first require 2 or more functions to be executed that are independent of each other.

### ISSUE THREAD

**Issue by norhh** [ConsenSys/mythril#1141]
## Description
For the contract below mythril doesn't detect this with dependency pruning enabled.
```


contract Rubixi {
    uint private collectedFees = 0;

    address private creator;


    function() external payable {
      if (msg.value < 1 ether ) {
            collectedFees += msg.value;
            return;
      }

    }

    function dynamicPyramid() public {
        creator = msg.sender;
    }


    function collectAllFees() public{
        require(collectedFees > 0);
        creator.transfer(collectedFees);
        collectedFees = 0;
    }

```

## How to Reproduce

```console
$ myth a file.sol
nothing
...
$
```

## Expected behavior
It should be detected

## Environment
Please modify for your setup

- Mythril version: `develop`
- Solidity compiler and version: `0.4.24`
- Python version: `3.6.7`
- OS and Version: Ubuntu 18.10

**Comment by b-mueller** [comment]
Well damn.

The fallback function and the `dynamicPyramid()` aren't dependent so with the plugin enabled, Mythril will not execute `fallback` followed by `dynamicPyramid()` or `dynamicPyramid()` followed by `fallback` :(.

**Comment by b-mueller** [comment]
This works though:

```
contract Bla {
    uint a = 0;
    uint b = 0;

    function aa() public {
        a = 1;
    }

    function bb() public {
        b = 1;
    }

    function x() public{
        require(a == 1);
        require(b == 1);
        selfdestruct(msg.sender);
    }
}
```

Not exactly sure what's the difference.

**Comment by b-mueller** [comment]
This works though:

```
contract Bla {
    uint private a = 0;
    uint private b = 0;

    function() external payable {
      if (msg.value < 1 ether ) {
            a = msg.value;
            return;
      }

    }

    function _b() public {
        b = 1;
    }

    function collectAllFees() public{
        require(a == 1);
        require(b == 1);
        selfdestruct(msg.sender);
    }
}
```

Not sure what's the fundamental difference.

**Comment by b-mueller** [comment]
This works:

```
contract Bla {
    uint private a = 0;
    uint private b = 0;
    address payable receiver;

    function _a() external payable {
      if (msg.value < 1 ether ) {
            a += msg.value;
            return;
      }

    }

    function _b() public {
        // receiver = msg.sender;
        b = 1;
    }

    function collectAllFees() public{
        require(a == 1);
        require(b == 1);
        selfdestruct(msg.sender);
    }
}
```


This doesn't work:

```
contract Bla {
    uint private a = 0;
    uint private b = 0;
    address payable receiver;

    function _a() external payable {
      if (msg.value < 1 ether ) {
            a += msg.value;
            return;
      }

    }

    function _b() public {
        receiver = msg.sender;
        // b = 1;
    }

    function collectAllFees() public{
        require(a == 1);
        require(receiver == msg.sender);
        selfdestruct(msg.sender);
    }
}
```

**Comment by norhh** [comment]
seems to be some ordering issue of functions. if we try 4 transactions these seem to work.

**Comment by b-mueller** [comment]
Yeah, it seems like.

This also works:

```
contract Bla {
    uint private a = 0;
    address payable receiver;

    function _a() public {
      receiver = msg.sender;
    }

    function _b() external payable {
        a = msg.value;
    }

    function collectAllFees() public {
        require(a == 1);
        receiver.transfer(a);
    }
}
```

**Comment by b-mueller** [comment]
And another fun observation! It works with:

```
        a = msg.value;
```

But not:

```
        a += msg.value;
```

For some reason the SLOAD makes a difference.

**Comment by b-mueller** [comment]
OK, I get it now. Ironically, if the plugin would work correctly it should *always* prune the path and the issue would never be detected (it's actually a "pure" path from the view of the plugin because it doesn't depend on the value of any state variables).

But with `a = msg.value` there is SLOAD and no entry in the dependency map is created. In it that case the plugin defaults to "don't prune" which "saves" the path.

In summary it's not great :(

**Comment by b-mueller** [comment]
Fixed by #1146
